pkgs/tools/nix/nixos-render-docs/src/tests/test_redirects.py

```diff
@@ -0,0 +1,228 @@
+        )
+        self.run_test(before)
+
+        intermediate = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["foo.html"]},
+        )
+        self.assert_redirect_error({"identifiers_missing_current_outpath": ["bar"]}, intermediate)
+
+        after = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["bar.html", "foo.html#bar"]},
+        )
+        self.run_test(after)
+
+    def test_non_leaf_identifier_moved_to_different_file(self):
+        """Test moving a non-leaf identifier to a different output path."""
+        before = self.setup_test(
+            sources={"foo.md": "# Foo {#foo}\n## Bar {#bar}\n### Baz {#baz}"},
+            raw_redirects={"foo": ["foo.html"], "bar": ["foo.html"], "baz": ["foo.html"]},
+        )
+        self.run_test(before)
+
+        intermediate = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}\n## Baz {#baz}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["foo.html"], "baz": ["foo.html"]},
+        )
+        self.assert_redirect_error({"identifiers_missing_current_outpath": ["bar", "baz"]}, intermediate)
+
+        after = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}\n## Baz {#baz}"
+            },
+            raw_redirects={"foo": ["foo.html"], "bar": ["bar.html", "foo.html#bar"], "baz": ["bar.html", "foo.html#baz"]},
+        )
+        self.run_test(after)
+
+    def test_conflicting_anchors(self):
+        """Test for conflicting anchors."""
+        md = self.setup_test(
+            sources={"foo.md": "# Foo {#foo}\n## Bar {#bar}"},
+            raw_redirects={
+                "foo": ["foo.html", "foo.html#bar"],
+                "bar": ["foo.html"],
+            }
+        )
+        self.assert_redirect_error({"conflicting_anchors": ["bar"]}, md)
+
+    def test_divergent_redirect(self):
+        """Test for divergent redirects."""
+        md = self.setup_test(
+            sources={
+                "foo.md": "# Foo {#foo}",
+                "bar.md": "# Bar {#bar}"
+            },
+            raw_redirects={
+                "foo": ["foo.html", "old-foo.html"],
+                "bar": ["bar.html", "old-foo.html"]
+            }
+        )
+        self.assert_redirect_error({"divergent_redirects": ["old-foo.html"]}, md)
+
+    def test_client_path_with_server_redirect(self):
+        """Test for client paths with server redirects."""
+        md = self.setup_test(
+            sources={"foo.md": "# Foo {#foo}"},
+            raw_redirects={"foo": ["foo.html", "bar.html", "bar.html#foo"]}
```

Comment:
Also I don't think that with the redirects checks enabled, we'll ever have a situation where a file will move as a whole merely by moving identifiers. Even top-level headings will require creating such an element-wise entry.

Reply:
> What we'd want in that situation is a new server-side redirect bar.html -> foo.html and that requires rewriting all bar.html entries in the redirects database to foo.html because otherwise those redirects just won't apply for the new output file.

That's precisely what I'm enforcing with this rule.

> We could catch such a situation programmatically, in the special case where the elements in the new location are a superset of the elements in the old location.

That's certainly possible, with little effort too I think. It should be as simple as modifying `get_client_redirects()` to check if its `redirection_target` exists as a destination path of some server-side redirect, and pull in all the client-side redirects from its the source path.

Should we do it this way? It would lead to smaller diffs when server-side redirects ever happen. Not sure what are the cons of doing this.

Reply:
> That's certainly possible, with little effort too I think. It should be as simple as modifying get_client_redirects() to check if its redirection_target exists as a destination path of some server-side redirect, and pull in all the client-side redirects that originate from its the source path.

Yes, sounds about right. Again, this is for a follow-up.